File: sdks/typescript/tests/unit/hash.test.ts
/**
 * Unit tests for hash functions.
 *
 * CRITICAL: These tests validate that our hash functions produce correct results.
 * Any failures here could indicate routing bugs that would cause data loss.
 */

import { describe, it, expect, beforeAll } from 'vitest';
import {
  initializeHasher,
  xxhash64,
  jumpConsistentHash,
  getShardForKey,
  keyToBytes,
  valueToBytes,
  bytesToString,
} from '@norikv/client/hash';

describe('Hash Functions', () => {
  beforeAll(async () => {
    await initializeHasher();
  });

  describe('xxhash64', () => {
    it('should hash strings consistently', async () => {
      const hash1 = xxhash64('hello');
      const hash2 = xxhash64('hello');
      expect(hash1).toBe(hash2);
    });

    it('should hash Uint8Array consistently', async () => {
      const bytes = new TextEncoder().encode('hello');
      const hash1 = xxhash64(bytes);
      const hash2 = xxhash64(bytes);
      expect(hash1).toBe(hash2);
    });

    it('should produce same hash for string and Uint8Array', async () => {
      const str = 'hello';
      const bytes = new TextEncoder().encode(str);
      const hash1 = xxhash64(str);
      const hash2 = xxhash64(bytes);
      expect(hash1).toBe(hash2);
    });

    it('should produce different hashes for different inputs', async () => {
      const hash1 = xxhash64('hello');
      const hash2 = xxhash64('world');
      expect(hash1).not.toBe(hash2);
    });

    it('should produce 64-bit values', async () => {
      const hash = xxhash64('test');
      expect(typeof hash).toBe('bigint');
      // Ensure it fits in 64 bits
      expect(hash).toBeGreaterThanOrEqual(0n);
      expect(hash).toBeLessThan(2n ** 64n);
    });
  });

  describe('jumpConsistentHash', () => {
    it('should return consistent bucket for same hash', () => {
      const hash = 12345678901234567890n;
      const bucket1 = jumpConsistentHash(hash, 1024);
      const bucket2 = jumpConsistentHash(hash, 1024);
      expect(bucket1).toBe(bucket2);
    });

    it('should return bucket in valid range', () => {
      const hash = xxhash64('test-key');
      const numBuckets = 1024;
      const bucket = jumpConsistentHash(hash, numBuckets);
      expect(bucket).toBeGreaterThanOrEqual(0);
      expect(bucket).toBeLessThan(numBuckets);
    });

    it('should distribute keys across buckets', () => {
      const buckets = new Set<number>();
      const numBuckets = 100;

      for (let i = 0; i < 1000; i++) {
        const key = `key-${i}`;
        const hash = xxhash64(key);
        const bucket = jumpConsistentHash(hash, numBuckets);
        buckets.add(bucket);
      }

      // With 1000 keys and 100 buckets, we should hit most buckets
      expect(buckets.size).toBeGreaterThan(90); // At least 90% coverage
    });

    it('should minimize moves when adding buckets', () => {
      const keys = Array.from({ length: 1000 }, (_, i) => `key-${i}`);

      // Map keys to 100 buckets
      const mapping100 = new Map<string, number>();
      for (const key of keys) {
        const hash = xxhash64(key);
        mapping100.set(key, jumpConsistentHash(hash, 100));
      }

      // Map same keys to 101 buckets
      const mapping101 = new Map<string, number>();
      for (const key of keys) {
        const hash = xxhash64(key);
        mapping101.set(key, jumpConsistentHash(hash, 101));
      }

      // Count how many keys moved
      let moved = 0;
      for (const key of keys) {
        if (mapping100.get(key) !== mapping101.get(key)) {
          moved++;
        }
      }

      // With consistent hashing, roughly 1/101 (~10) keys should move
      // Allow some variance: between 1 and 20 keys
      expect(moved).toBeGreaterThan(0);
      expect(moved).toBeLessThan(20);
    });

    it('should throw on invalid bucket count', () => {
      const hash = 123n;
      expect(() => jumpConsistentHash(hash, 0)).toThrow();
      expect(() => jumpConsistentHash(hash, -1)).toThrow();
    });
  });

  describe('getShardForKey', () => {
    it('should return consistent shard for same key', async () => {
      const shard1 = getShardForKey('user:123', 1024);
      const shard2 = getShardForKey('user:123', 1024);
      expect(shard1).toBe(shard2);
    });

    it('should return shard in valid range', async () => {
      const shard = getShardForKey('test-key', 1024);
      expect(shard).toBeGreaterThanOrEqual(0);
      expect(shard).toBeLessThan(1024);
    });

    it('should distribute keys across shards', async () => {
      const shards = new Set<number>();

      for (let i = 0; i < 10000; i++) {
        const key = `key-${i}`;
        const shard = getShardForKey(key, 1024);
        shards.add(shard);
      }

      // With 10k keys and 1024 shards, we should hit most shards
      expect(shards.size).toBeGreaterThan(1000); // At least 97% coverage
    });
  });

  describe('keyToBytes / valueToBytes', () => {
    it('should convert string to bytes', () => {
      const str = 'hello';
      const bytes = keyToBytes(str);
      expect(bytes).toBeInstanceOf(Uint8Array);
      expect(bytesToString(bytes)).toBe(str);
    });

    it('should pass through Uint8Array unchanged', () => {
      const original = new Uint8Array([1, 2, 3, 4, 5]);
      const bytes = keyToBytes(original);
      expect(bytes).toBe(original); // Same reference
    });

    it('should handle UTF-8 strings correctly', () => {
      const str = 'Hello 世界 🌍';
      const bytes = keyToBytes(str);
      expect(bytesToString(bytes)).toBe(str);
    });

    it('should handle empty strings', () => {
      const bytes = keyToBytes('');
      expect(bytes.length).toBe(0);
      expect(bytesToString(bytes)).toBe('');
    });
  });

  describe('Cross-language compatibility test vectors', () => {
    // These test vectors should match the server's Rust implementation
    // TODO: Generate these from the server and commit them to version control

    it('should match known hash values', async () => {
      // Test vector: "hello" with seed=0
      // This value should be verified against the Rust implementation
      const hash = xxhash64('hello');
      expect(typeof hash).toBe('bigint');
      // TODO: Replace with actual expected value from server
      // expect(hash).toBe(EXPECTED_HASH_FROM_SERVER);
    });

    it('should match known shard assignments', async () => {
      // Test vector: "user:123" -> shard 42 (example)
      const shard = getShardForKey('user:123', 1024);
      expect(shard).toBeGreaterThanOrEqual(0);
      expect(shard).toBeLessThan(1024);
      // TODO: Replace with actual expected shard from server
      // expect(shard).toBe(EXPECTED_SHARD_FROM_SERVER);
    });
  });

  describe('Performance benchmarks', () => {
    it('should hash quickly', async () => {
      const iterations = 10000;
      const start = Date.now();

      for (let i = 0; i < iterations; i++) {
        xxhash64(`key-${i}`);
      }

      const elapsed = Date.now() - start;
      const perOp = elapsed / iterations;

      // Should be well under 1ms per operation
      expect(perOp).toBeLessThan(1);
    });

    it('should compute shard quickly', async () => {
      const iterations = 10000;
      const start = Date.now();

      for (let i = 0; i < iterations; i++) {
        getShardForKey(`key-${i}`, 1024);
      }

      const elapsed = Date.now() - start;
      const perOp = elapsed / iterations;

      // Should be well under 1ms per operation
      expect(perOp).toBeLessThan(1);
    });
  });
});

File: sdks/go/hash/hash_test.go
package hash

import (
	"testing"
)

func TestXXHash64Consistency(t *testing.T) {
	t.Run("hash strings consistently", func(t *testing.T) {
		key := []byte("hello")
		hash1 := XXHash64(key)
		hash2 := XXHash64(key)
		if hash1 != hash2 {
			t.Errorf("Inconsistent hashes: %d != %d", hash1, hash2)
		}
	})

	t.Run("different inputs produce different hashes", func(t *testing.T) {
		hash1 := XXHash64([]byte("hello"))
		hash2 := XXHash64([]byte("world"))
		if hash1 == hash2 {
			t.Errorf("Different inputs produced same hash: %d", hash1)
		}
	})

	t.Run("handles empty string", func(t *testing.T) {
		hash := XXHash64([]byte(""))
		// Just verify it doesn't panic and returns a value
		if hash < 0 {
			t.Errorf("Hash should be non-negative, got: %d", hash)
		}
	})

	t.Run("handles unicode", func(t *testing.T) {
		hash := XXHash64([]byte("Hello 世界 🌍"))
		if hash < 0 {
			t.Errorf("Hash should be non-negative, got: %d", hash)
		}
	})

	t.Run("handles large keys", func(t *testing.T) {
		largeKey := make([]byte, 10000)
		for i := range largeKey {
			largeKey[i] = 'x'
		}
		hash := XXHash64(largeKey)
		if hash < 0 {
			t.Errorf("Hash should be non-negative, got: %d", hash)
		}
	})
}

func TestJumpConsistentHash(t *testing.T) {
	t.Run("consistent bucket for same hash", func(t *testing.T) {
		hash := uint64(12345678901234567890)
		bucket1 := JumpConsistentHash(hash, 1024)
		bucket2 := JumpConsistentHash(hash, 1024)
		if bucket1 != bucket2 {
			t.Errorf("Inconsistent buckets: %d != %d", bucket1, bucket2)
		}
	})

	t.Run("bucket in valid range", func(t *testing.T) {
		hash := XXHash64([]byte("test-key"))
		numBuckets := 1024
		bucket := JumpConsistentHash(hash, numBuckets)
		if bucket < 0 || bucket >= numBuckets {
			t.Errorf("Bucket %d out of range [0, %d)", bucket, numBuckets)
		}
	})

	t.Run("distributes keys across buckets", func(t *testing.T) {
		buckets := make(map[int]bool)
		numBuckets := 100

		for i := 0; i < 1000; i++ {
			key := []byte("key-" + string(rune(i)))
			hash := XXHash64(key)
			bucket := JumpConsistentHash(hash, numBuckets)
			buckets[bucket] = true
		}

		// With 1000 keys and 100 buckets, should hit most buckets
		if len(buckets) <= 90 {
			t.Errorf("Poor distribution: only %d/%d buckets used", len(buckets), numBuckets)
		}
	})

	t.Run("minimizes moves when adding buckets", func(t *testing.T) {
		numKeys := 1000
		keys := make([][]byte, numKeys)
		for i := 0; i < numKeys; i++ {
			keys[i] = []byte("key-" + string(rune(i)))
		}

		// Map keys to 100 buckets
		mapping100 := make(map[string]int)
		for _, key := range keys {
			hash := XXHash64(key)
			mapping100[string(key)] = JumpConsistentHash(hash, 100)
		}

		// Map same keys to 101 buckets
		mapping101 := make(map[string]int)
		for _, key := range keys {
			hash := XXHash64(key)
			mapping101[string(key)] = JumpConsistentHash(hash, 101)
		}

		// Count how many keys moved
		moved := 0
		for _, key := range keys {
			if mapping100[string(key)] != mapping101[string(key)] {
				moved++
			}
		}

		// With consistent hashing, roughly 1/101 (~10) keys should move
		// Allow some variance: between 1 and 20 keys
		if moved <= 0 || moved >= 20 {
			t.Errorf("Too many keys moved: %d (expected 1-20)", moved)
		}
	})

	t.Run("panics on invalid bucket count", func(t *testing.T) {
		defer func() {
			if r := recover(); r == nil {
				t.Errorf("Expected panic for invalid bucket count")
			}
		}()
		JumpConsistentHash(123, 0)
	})

	t.Run("handles single bucket", func(t *testing.T) {
		hash := XXHash64([]byte("test"))
		bucket := JumpConsistentHash(hash, 1)
		if bucket != 0 {
			t.Errorf("Single bucket should be 0, got: %d", bucket)
		}
	})

	t.Run("handles large bucket count", func(t *testing.T) {
		hash := XXHash64([]byte("test"))
		numBuckets := 1000000
		bucket := JumpConsistentHash(hash, numBuckets)
		if bucket < 0 || bucket >= numBuckets {
			t.Errorf("Bucket %d out of range [0, %d)", bucket, numBuckets)
		}
	})
}

func TestGetShardForKey(t *testing.T) {
	t.Run("consistent shard for same key", func(t *testing.T) {
		shard1 := GetShardForKey([]byte("user:123"), 1024)
		shard2 := GetShardForKey([]byte("user:123"), 1024)
		if shard1 != shard2 {
			t.Errorf("Inconsistent shards: %d != %d", shard1, shard2)
		}
	})

	t.Run("shard in valid range", func(t *testing.T) {
		shard := GetShardForKey([]byte("test-key"), 1024)
		if shard < 0 || shard >= 1024 {
			t.Errorf("Shard %d out of range [0, 1024)", shard)
		}
	})

	t.Run("distributes keys across shards", func(t *testing.T) {
		shards := make(map[int]bool)

		for i := 0; i < 10000; i++ {
			key := []byte("key-" + string(rune(i)))
			shard := GetShardForKey(key, 1024)
			shards[shard] = true
		}

		// With 10k keys and 1024 shards, should hit most shards
		if len(shards) <= 1000 {
			t.Errorf("Poor distribution: only %d/1024 shards used", len(shards))
		}
	})
}

// TestCrossSDKCompatibility validates parity with Python and TypeScript SDKs.
// CRITICAL: These test vectors must match exactly or routing will fail.
func TestCrossSDKCompatibility(t *testing.T) {
	t.Run("known hash values", func(t *testing.T) {
		// Test vectors generated from TypeScript SDK
		// Generated by sdks/typescript/scripts/generate-hash-vectors.ts
		// These MUST match exactly
		testVectors := []struct {
			key          string
			expectedHash uint64
		}{
			{"hello", 2794345569481354659},
			{"world", 16679358290033791471},
			{"user:123", 12838382303697200627},
			{"test-key", 15827481653422919556},
			{"", 17241709254077376921},
			{"🌍", 11345255911858566016},
			{"foo", 3728699739546630719},
			{"bar", 5234164152756840025},
			{"a", 15154266338359012955},
			{"ab", 7347350983217793633},
			{"abc", 4952883123889572249},
			{"test-value-1", 2906801998163794869},
			{"test-value-2", 2618596406459050154},
		}

		for _, tv := range testVectors {
			actualHash := XXHash64([]byte(tv.key))
			if actualHash != tv.expectedHash {
				t.Errorf("Hash mismatch for '%s': expected %d, got %d",
					tv.key, tv.expectedHash, actualHash)
			}
		}
	})

	t.Run("known shard assignments", func(t *testing.T) {
		// Test vectors generated from TypeScript SDK
		// Generated by sdks/typescript/scripts/generate-hash-vectors.ts
		// These MUST match exactly
		testVectors := []struct {
			key           string
			totalShards   int
			expectedShard int
		}{
			{"hello", 1024, 309},
			{"world", 1024, 752},
			{"user:123", 1024, 928},
			{"test-key", 1024, 504},
			{"", 1024, 332},
			{"🌍", 1024, 393},
			{"foo", 1024, 951},
			{"bar", 1024, 632},
			{"a", 1024, 894},
			{"ab", 1024, 335},
			{"abc", 1024, 722},
			{"test-value-1", 1024, 640},
			{"test-value-2", 1024, 488},
		}

		for _, tv := range testVectors {
			actualShard := GetShardForKey([]byte(tv.key), tv.totalShards)
			if actualShard != tv.expectedShard {
				t.Errorf("Shard mismatch for '%s': expected %d, got %d",
					tv.key, tv.expectedShard, actualShard)
			}
		}
	})

	t.Run("jump consistent hash values", func(t *testing.T) {
		// Test vectors for jump_consistent_hash algorithm
		// Generated by sdks/typescript/scripts/generate-hash-vectors.ts
		testVectors := []struct {
			hashValue      uint64
			numBuckets     int
			expectedBucket int
		}{
			{0, 100, 0},
			{1, 100, 55},
			{12345678901234567890, 1024, 294},
			{0xFFFFFFFFFFFFFFFF, 1024, 313},
		}

		for _, tv := range testVectors {
			actualBucket := JumpConsistentHash(tv.hashValue, tv.numBuckets)
			if actualBucket != tv.expectedBucket {
				t.Errorf("JCH mismatch for hash=%d, buckets=%d: expected %d, got %d",
					tv.hashValue, tv.numBuckets, tv.expectedBucket, actualBucket)
			}
		}
	})
}

// TestKeyValueConversions tests convenience conversion functions.
func TestKeyValueConversions(t *testing.T) {
	t.Run("KeyToBytes", func(t *testing.T) {
		key := "hello"
		bytes := KeyToBytes(key)
		if string(bytes) != key {
			t.Errorf("KeyToBytes mismatch: expected %s, got %s", key, string(bytes))
		}
	})

	t.Run("ValueToBytes", func(t *testing.T) {
		value := "world"
		bytes := ValueToBytes(value)
		if string(bytes) != value {
			t.Errorf("ValueToBytes mismatch: expected %s, got %s", value, string(bytes))
		}
	})

	t.Run("BytesToString", func(t *testing.T) {
		bytes := []byte("hello")
		str := BytesToString(bytes)
		if str != "hello" {
			t.Errorf("BytesToString mismatch: expected hello, got %s", str)
		}
	})

	t.Run("UTF-8 round trip", func(t *testing.T) {
		original := "Hello 世界 🌍"
		bytes := KeyToBytes(original)
		result := BytesToString(bytes)
		if result != original {
			t.Errorf("UTF-8 round trip failed: expected %s, got %s", original, result)
		}
	})

	t.Run("empty string", func(t *testing.T) {
		bytes := KeyToBytes("")
		if len(bytes) != 0 {
			t.Errorf("Empty string should produce empty bytes, got length %d", len(bytes))
		}
		str := BytesToString(bytes)
		if str != "" {
			t.Errorf("Empty bytes should produce empty string, got %s", str)
		}
	})
}

// BenchmarkXXHash64 benchmarks the hash function performance.
func BenchmarkXXHash64(b *testing.B) {
	key := []byte("test-key")
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		XXHash64(key)
	}
}

// BenchmarkGetShardForKey benchmarks the shard assignment performance.
func BenchmarkGetShardForKey(b *testing.B) {
	key := []byte("test-key")
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		GetShardForKey(key, 1024)
	}
}

// BenchmarkJumpConsistentHash benchmarks the consistent hash algorithm.
func BenchmarkJumpConsistentHash(b *testing.B) {
	hash := uint64(12345678901234567890)
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		JumpConsistentHash(hash, 1024)
	}
}
